### PR TITLE
OpenID with refreshToken to refresh the access_token

### PR DESCRIPTION
**Pull request by agarwalgovind**
*Issue #, if available:*

While using OpenID with external Idp like keycloak etc, the access token from the IDP doesnt get refreshed and the Kibana page get struck or the user will be redirected to the login page.

*Description of changes:*

Used few of the changes done as per the existing PR https://github.com/opendistro-for-elasticsearch/security-kibana-plugin/pull/72/commits and made certain changes on top of it.

Changed the OpenID.js files to override methods of AuthType.js and implemented the logic to refresh the access token and updating the opendistro cookies accordingly and keeping the session active.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by skleis** [comment]
Hi, I am no expert in Typescript or Kibana plugin development, but it seams that the you are storing the refesh_token_object, irrespective of any user-session in the Kibana backend. This would be a major security risk and should only be used, when running Kibana in a local instance where only one user has access to the Kibana. 

Am I correct in this assumption or do I miss something?

**Comment by agarwalgovind** [comment]
Hi @skleis I guess you are correct, i missed this part. Thanks for pointing it out.

I guess i should store it as a map in Kibana backend with Key as session-id. Can you suggest your views or alternative solution ?

**Comment by skleis** [comment]
Hi @agarwalgovind 

Storing the refresh tokens in a map would work for single running Kibana instance, but would fail for a multi-instance use-case. I'm currently evaluating the possibility of storing the refresh token in the securitySession itself, where it would be shareble but encrypted. 

I need a bit more time to understand the dataflow/functionality of the openid-auth implementation to be able to extend the session storage with refresh token and if there are no hidden vulrablities in this solution.

**Comment by agarwalgovind** [comment]
Hi @skleis,

I could not get the issue with multi instance of kibana. If there are multi instance of kibana then each instance will have its seperate in-memory as well as security-session also. So even if we store refresh_token in securitySession of one instance, i dont think it will be available to the other.

Hpwever for multi instance Kibana running behind a Load Balancer I am still trying to understand how kibana instances will run and how it will support HA mode also

**Comment by skleis** [comment]
Hi @agarwalgovind,

I found a solution that works no with multi instance kibana. The method of storing the refresh token in the session worked perfectly. The securitySession is not stored on the Kibana server but rather as an encrypted cookie in the browser. 

Some caveats:
1. A refersh_token is stored in the browser, which is against Oauth2 best practices. This is mitigated by the encrpytion so the token cannot be used outside of the kibana instance. Furthermore for our use-case the refresh_token stored in the cookie has a short-ish lifetime, so a stolen session cookie has limited security risks.
2. The kibana session and cookies ttl ahve to be configured correctly to allow the refresh to function properly. The ttl for the session has to be lower than the ttl of the cookie, so that the refresh token can be correctly read from the encrypted cookie. In fact we set the ttl of the cookie to 0 to always have access to the refresh_token until the browser is closed.

I would like to share some code, but I have to check with the head of my department if I am allowed to do that.

**Comment by agarwalgovind** [comment]
Hi @skleis 

Thanks for the info, I am able to make the changes of storing the token in the securitySession cookie and the changes are woking now. I will raise the latest PR.

File: public/chrome/configuration/enable_configuration.js [hunk 1]
```diff
@@ -76,9 +76,17 @@ app.factory('errorInterceptor', function ($q, $window) {
 
     return {
         responseError: function (response) {
-
             // Handles 401s, but only if we've explicitly set the redirect property on the response.
-            if (response.status == 401 && response.data && response.data.redirectTo === 'login') {
+            // Fix for https://github.com/angular/angular/issues/19888
+            var data = null;
```
Comment (seraphjiang) [inline public/chrome/configuration/enable_configuration.js:81]: would you use let instead of var here

File: lib/auth/types/openid/routes.js [hunk 2]
```diff
@@ -109,29 +111,34 @@ module.exports = async function (pluginRoot, server, kbnServer, APP_ROOT, API_RO
...
-                    // Bell gives us the access token to identify with here,
-                    // but we want the id_token returned from the IDP
-                    let {user} = await request.auth.securitySessionStorage.authenticate({
-                        authHeaderValue: 'Bearer ' + request.auth.artifacts['id_token']
+                    // Bell gives us the access token to identify with here, but we want the id_token returned from the IDP
+                    if (credentials) {
+                        let {user} = await request.auth.securitySessionStorage.authenticate({
+                            authHeaderValue: 'Bearer ' + request.auth.artifacts['id_token'],
+                            refresh_token: request.auth.credentials.refreshToken
                     });
```
Comment (seraphjiang) [inline lib/auth/types/openid/routes.js:137]: would you fix the indentation

File: lib/auth/types/AuthType.js [hunk 3]
```diff
@@ -201,7 +210,7 @@ export default class AuthType {
                 try {
                     credentials = await this.server.auth.test('security_access_control_cookie', request);
                     return h.authenticated({credentials})
-                 } catch(error) {
+                } catch(error) {
```
Comment (seraphjiang) [inline lib/auth/types/AuthType.js:213]: would you help to revert the code formation change, it will help understand what's the actual change.

File: lib/auth/types/openid/OpenId.js [hunk 4]
```diff
@@ -67,13 +76,92 @@ export default class OpenId extends AuthType {
...
+                    /*Commented to stop Kibana getting struck as CORS policy comes into effect as it stops any outgoing kibana request to keycloak.
+                      On invalid token or failed to update access_token using refresh_token the code will reload Kibana with new token till IDP SSO
+                      is active. Once IDP SSO timeout the code will redirect to the login page for a fresh login.
+                    */
+
+                    /*if (this._cookieValidationError) {
+                      server.log(['error', 'security'], `Cookie Validation Error`);
+                      return this.onUnAuthenticated(request, h, this._cookieValidationError).takeover();
+                    }*/
+
```
Comment (seraphjiang) [inline lib/auth/types/openid/OpenId.js:116]: would you cleanup unused code

File: lib/auth/types/openid/OpenId.js [hunk 5]
```diff
@@ -67,13 +76,92 @@ export default class OpenId extends AuthType {
...
+
+                    let authHeaderCredentials = this.detectAuthHeaderCredentials(request);
+                    if (authHeaderCredentials) {
+                        try {
+                            let { session } = await request.auth.securitySessionStorage.authenticate(authHeaderCredentials);
```
Comment (seraphjiang) [inline lib/auth/types/openid/OpenId.js:121]: use const

File: lib/auth/types/openid/OpenId.js [hunk 6]
```diff
@@ -111,39 +199,158 @@ export default class OpenId extends AuthType {
...
             return h.redirect(this.basePath + '/customerror?type=missingTenant')
         } else if (error instanceof MissingRoleError) {
             return h.redirect(this.basePath + '/customerror?type=missingRole')
         } else if (error instanceof AuthenticationError) {
+            this.server.log(['error', 'security'], `Error occured with authentication: ${error}`);
             return h.redirect(this.basePath + '/customerror?type=authError')
         }
 
-        const nextUrl = this.getNextUrl(request);
+        //this.server.log(['error', 'security'], `onAuthenticated - Redirect`);
```
Comment (seraphjiang) [inline lib/auth/types/openid/OpenId.js:212]: cleanup unused code

File: lib/auth/types/openid/OpenId.js (hunk 4 above)
Comment (agarwalgovind) [reply lib/auth/types/openid/OpenId.js]: Done

File: lib/auth/types/openid/OpenId.js (hunk 6 above)
Comment (agarwalgovind) [reply lib/auth/types/openid/OpenId.js]: Done

File: lib/auth/types/openid/OpenId.js (hunk 5 above)
Comment (agarwalgovind) [reply lib/auth/types/openid/OpenId.js]: Done

File: lib/auth/types/AuthType.js [hunk 7]
```diff
@@ -244,10 +253,10 @@ export default class AuthType {
         this.server.auth.strategy('security_access_control', 'security_access_control_scheme', this.getCookieConfig());
         this.server.auth.strategy('security_access_control_cookie', 'cookie', this.getCookieConfig());
 
-         this.server.auth.default({
+        this.server.auth.default({
             mode: 'required', // @todo Investigate best mode here
             strategy: 'security_access_control' // This seems to be the only way to apply the strategy to ALL routes, even those defined before we add the strategy.
-         });
+        });
```
Comment (palashhedau) [inline lib/auth/types/AuthType.js:259]: I assume no change before this apart from "routesToIgnore"?

File: lib/auth/types/openid/OpenId.js [hunk 8]
```diff
@@ -73,7 +80,72 @@ export default class OpenId extends AuthType {
...
+                if (this.unauthenticatedRoutes.includes(request.path)) {
+                    credentials = this.server.plugins.opendistro_security.getSecurityBackend().getServerUser();
+                    return h.authenticated({ credentials });
+                };
+
+                try {
+                    credentials = await this.server.auth.test('security_access_control_cookie', request);
+                    return h.authenticated({ credentials })
+                } catch (error) {
+                    let authHeaderCredentials = this.detectAuthHeaderCredentials(request);
```
Comment (palashhedau) [inline lib/auth/types/openid/OpenId.js:103]: Are we not doing anything for "_cookieValidationError" like in superclass?

File: lib/auth/types/openid/OpenId.js [hunk 9]
```diff
@@ -121,29 +193,141 @@ export default class OpenId extends AuthType {
             return h.redirect(this.basePath + '/customerror?type=authError')
         }
 
-        const nextUrl = this.getNextUrl(request);
+        const nextUrl = encodeURIComponent(request.url.path);
```
Comment (palashhedau) [inline lib/auth/types/openid/OpenId.js:196]: What about request.getBasePath()?

File: lib/auth/types/openid/OpenId.js [hunk 10]
```diff
@@ -121,29 +193,141 @@ export default class OpenId extends AuthType {
...
+                        let body = await this.refreshTokenRequest(session.credentials.refresh_token);
+                        let result = JSON.parse(body);
+                        let id_token = result['id_token'];
+                        let tokenPayload = JSON.parse(Buffer.from(id_token.split('.')[1], 'base64').toString());
+                        refreshedSession.username = tokenPayload['name'];
+                        refreshedSession.credentials = {};
+                        refreshedSession.credentials.authHeaderValue = 'Bearer ' + id_token;
+                        refreshedSession.authType = 'openid';
+                        if (tokenPayload.exp) {
+                            this.sessionKeepAlive = false;
```
Comment (palashhedau) [inline lib/auth/types/openid/OpenId.js:284]: shouldn't this be true?

Comment (palashhedau) [inline lib/auth/types/openid/OpenId.js:284]: Are we using sessionKeepAlive in openid anywhere?

File: lib/auth/types/openid/OpenId.js [hunk 11]
```diff
@@ -121,29 +193,141 @@ export default class OpenId extends AuthType {
...
+                        return { valid: false };
+                    }
+                } else {
+                        this.server.log(['error', 'security'], `Error while fetching refreshToken : ${error}`);
+                        request.auth.securitySessionStorage.clearStorage();
+                        return { valid: false };
+                    }
+            } else if (!session.exp && this.sessionTTL) {
+                if (!session.expiryTime || session.expiryTime < Date.now()) {
+                    request.auth.securitySessionStorage.clearStorage();
```
Comment (palashhedau) [inline lib/auth/types/openid/OpenId.js:305]: Why is this removed ? this._cookieValidationError = new SessionExpiredError('Session expired')

File: lib/auth/types/openid/routes.js [hunk 12]
```diff
@@ -129,9 +131,15 @@ module.exports = async function (pluginRoot, server, kbnServer, APP_ROOT, API_RO
                 try {
                     // Bell gives us the access token to identify with here,
                     // but we want the id_token returned from the IDP
-                    let {user} = await request.auth.securitySessionStorage.authenticate({
-                        authHeaderValue: 'Bearer ' + request.auth.artifacts['id_token']
-                    });
+                    if (credentials) {
+                        let {user} = await request.auth.securitySessionStorage.authenticate({
+                            authHeaderValue: 'Bearer ' + request.auth.artifacts['id_token'],
+                            refresh_token: request.auth.credentials.refreshToken
```
Comment (palashhedau) [inline lib/auth/types/openid/routes.js:137]: Will the refresh token always here even when login for the first time?

**Review by palashhedau — COMMENTED** [review]
Most of the overridden function are same with minor changes. 
Please mention those changes function-wise in description if possible

File: lib/auth/types/AuthType.js (hunk 7 above)
Comment (agarwalgovind) [reply lib/auth/types/AuthType.js:259]: Yes

File: lib/auth/types/openid/OpenId.js (hunk 8 above)
Comment (agarwalgovind) [reply lib/auth/types/openid/OpenId.js:103]: _cookieValidationError is removed as the code to check for the error is removed.

With respect to superclass there is a considerable change for handling the refresh_token.

File: lib/auth/types/openid/OpenId.js (hunk 11 above)
Comment (agarwalgovind) [reply lib/auth/types/openid/OpenId.js:305]: _cookieValidationError is removed as the code to check for the error is removed. For openId, the check to handle cookieValidation is redirecting the user for a login, but when using with IDP the request is not redirected to IDP as Kibana getting struck as CORS policy comes into effect as it stops any outgoing kibana request to IDP.

File: lib/auth/types/openid/routes.js (hunk 12 above)
Comment (agarwalgovind) [reply lib/auth/types/openid/routes.js:137]: Yes

**Review by palashhedau — APPROVED** [review]
Most of the overridden function are same with minor changes. 
Please mention those changes function-side in dexcription if possible

**Review by seraphjiang — APPROVED** [review]
